### ISSUE TITLE
Performance benchmark as part of CI/CD

### ISSUE THREAD

**Issue by scnakandala** [microsoft/hummingbird#271]
I was wondering whether we should have some kind of performance-based tests as part of the  CI/CD pipelines.
Any thoughts?

**Comment by interesaaat** [comment]
I was thinking about something similar today. It would be good to have some benchmarking refresh every time a new commit goes in. Perhaps we can store the benchmark numbers in the same location as the documentation. 

The only problem I see is that if we add benchmarks as part of the ci/cd, the pipelines will take forever. Maybe we can have not a full benchmark as we have in the paper but just some performance test as you are suggesting. Like a couple of tests per operator or something like that.

**Comment by ksaur** [comment]
I love the *idea* of performance-based tests, because it would be great to be auto-notified if there is a regression (especially when new versions of pytorch, etc, change things unexpectedly).

But the CI/CD is already _crawling_.  If we do add something to the pipeline, it would have to be minimal.   Matteo's idea of a  couple of tests only could work!

Regarding benchmarks in general,  one thought that crossed my mind is to post the artifact eval scripts we are polishing up for OSDI.   (Make them available but not part of pipeline directly)

**Comment by interesaaat** [comment]
Yes, we can put all the scripts into a `benchmark` directory. Do you know how other projects deals with performance regression in an (semi-)automatic way?

**Comment by ksaur** [comment]
Maybe we can use something from [this github blog](https://github.blog/2020-06-17-using-github-actions-for-mlops-data-science/) which shows runtime.  It's not a perfect fit for what we're looking for, but it does provide some useful features.

**Comment by interesaaat** [comment]
I am starting with pushing the code for the benchmarks. The tasks to complete this PR will be:
- [ ] Trees [#315]
- [ ] Operators [#316]
- [ ] Pipelines [#317]
- [ ] Add benchmarking to CI/CD pipeline
